woocommerce/woocommerce
# Refactor the settings pages, and add unit tests for them

**Note:** [The original pull request](https://github.com/woocommerce/woocommerce/pull/26024) has been closed because it was very old and there was no way to sensibly merge the changes from master, not because of merge conflicts (there were just a handful and not very complex) but because of issues with the code sniffer (it was complaining about one of the configured sniffs being unknown, and complaining about each and every .js file in the codebase).

### All Submissions:



### Changes proposed in this Pull Request:

This pull request introduces an extensive refactor of the WooCommerce settings classes (`WC_Settings_Page` and derived classes) that includes the following:

* Modify the `get_settings()` method so that it accepts a section id as a parameter, but instead of including this parameter in the method signature get it with `func_get_arg`

  * Some derived classes in the WooCommerce code had changed the method signature to `get_settings( $current_section = '' )` already, but others (those not having multiple sections natively) hadn't, making then effectively impossible to define multiple sections for these pages via filters.
  * `func_get_arg` is needed for compatibility with PHP 8 since some external derived classes may have redefined the signature to `get_settings( $current_section = '' )` but others will still be using `get_settings()`, see the comment added to the method for details.
  * Because of the described hack `get_settings` is marked as deprecated, `get_settings_for_section` should be used instead.
<br/>

* Add a `get_settings_for_section($section_id)` method in the base class that searches for a `get_settings_for_{section_id}_section` method in the class and executes it, otherwise it executes the new protected method `get_settings_for_section_core( $current_section )`; then it triggers the `'woocommerce_get_settings_' . id` filter.

  * This makes it easier to separate the code that returns the list of filters in multiple methods, one per section, instead of using one big if-else-else... block.
  * Also, `'woocommerce_get_settings_' . id` is triggered in one single place too.
  *  `get_settings_for_section` is marked as `final` because we want derived classes to either implement `get_settings_for_{section_id}_section` methods or override the `get_settings_for_section_core` method (or both).
<br/>

* General refactorings to fix code sniffer errors, including replacing old-style sniffer directives with phpcs:enable/disable/ignore comments.

* Some duplicated code has been refactored to protected methods.

* Changes the "todo: remove in 4.0" comment in `WC_Settings_Rest_API` to "remove in 5.0".


### How to manually test the changes in this Pull Request:

Create the following plugin and open the _General_ settings tab. You'll see that it has three sections: `General, Foobar, Fizzbuz`. The first one has one extra setting, the second one has just one setting, and the thirs one is empty. Without the changes introduced in this pull request you would have been able to create the sections, but the `woocommerce_get_settings_general` trigger would be invoked with only the `$settings` parameter and thus you wouldn't have been able to return different per section settings.

```php
<?php
/*
Plugin Name: Foobar & Fizzbuzz
Version: 1.0
Description: Foo? Bar. Fizz? Buzz. See? It's easy!
Author: WooCommerce (testing)
*/

add_filter('woocommerce_get_sections_general', 'foobar_sections_for_general_settings', 10, 1);

function foobar_sections_for_general_settings($settings) {
	$settings['foobars'] = 'Foobars';
	$settings['fizzbuzzs'] = 'Fizzbuzzs';
	return $settings;
}

add_filter('woocommerce_get_settings_general', 'foobar_settings_for_general_tab', 10, 2);

function foobar_settings_for_general_tab($settings, $current_section) {
	if( '' === $current_section ) {
		$settings[] =
			array(
				'title' => "Bonus settings",
				'type'  => 'title',
				'desc'  => "Some bonus settings inside the default section!"
			);

		$settings[] =
			array(
				'title'    => 'Some setting',
				'desc'     => 'Is this setting set or not?',
				'id'       => 'foobar_some_setting',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'A setting that can be set or not.'
			);

		$settings[] = array( 'type' => 'sectionend'	);

		return $settings;
	}
	else if( 'foobars' === $current_section ) {
		return array(
			array(
				'title' => "Foobars",
				'type'  => 'title',
				'desc'  => "What about foobars?"
			),

			array(
				'title'    => 'The foo is bar',
				'desc'     => 'States if the foo is bar',
				'id'       => 'foobar_the_foo_is_bar',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'Is the foo bar enough?',
			),

			array( 'type' => 'sectionend' )
		);
	}
	else {
		return array();
	}
}
```

It's also important to test with a settings class that directly overrides the `get_settings` method, [here's a handy code snippet for that](https://gist.github.com/claudiosanches/618892ba1c27c48e8298dab374d71216), or you can use an existing extension like [Freshbooks](https://github.com/woocommerce/woocommerce-freshbooks/blob/917a57b8851becd7dc62b39e99a198e4dd527d39/includes/admin/class-wc-freshbooks-settings.php).

### Other information:



### Changelog entry

> Enhancement - All settings pages can now be extended consistently with new sections and settings. Also, unit tests have been added.
